feat: acs endpoint

ilee2u opened the pull request:
**JIRA:** [Link to JIRA ticket](https://2u-internal.atlassian.net/browse/MST-1847)

**Description:** Adds an ACS endpoint to /lti/views.py

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Delete working branch (if not needed anymore)

Zacharis278 commented on edx_exams/apps/lti/views.py:
```diff
@@ -23,8 +28,53 @@
+    data = json.loads(request.body)
+    action = data['action']
+
+    if action == 'flag':
```
I'd recommend doing a bit more than just looking at the action here. We can take the user, resource, and attempt number in this request to look up the attempt object that we would be modifying or flagging. Can we throw some errors if those parameters are incorrect? Maybe make our log message more explicit on exactly what attempt is being flagged?

Zacharis278 commented on edx_exams/apps/lti/views.py:
```diff
@@ -23,8 +28,53 @@
+    elif action == 'resume':
+        log.info('Resuming exam attempt')
+
+    elif action == 'terminate':
```
since we're telling the tool we only support flag request I don't think we need conditions for all of the other possible actions yet (until we choose to support them)

Zacharis278 commented on edx_exams/apps/lti/urls.py:
```diff
@@ -8,6 +8,7 @@
 
 app_name = 'lti'
 urlpatterns = [
+    path('acs_endpoint/<int:lti_config_id>', views.acs_endpoint, name='acs_endpoint'),
```
NIT: 'endpoint' seems kinda redundant to put in a url. Maybe just acs or assessment_control. I'd also recommend putting the config id first in this url since that id refers to the lti configuration and not the assessment. IMO putting it last implies it's an assessment/attempt id.

so` /lti/<lti_configuration>/acs` or something like that

ilee2u commented on edx_exams/apps/lti/views.py:
```diff
@@ -12,19 +15,96 @@
+            f'with resource id {resource_id} and attempt',
+            f'number {attempt_number}.'
+        )
+        # NOTE: Do we want to create a new exception in exceptions.py just for this case?
```
Just want to highlight this question about making a new exception for this case.

MichaelRoytman replied:
You could, but how would you use it? You're checking the condition here, so this is where you'd raise the exception. But this is also the place you'd handle the exception.

I think this would be more useful if you were, say, performing a status transition in the API or something and needed to percolate up an exception to the calling code. We don't have a need for that yet, in my opinion.

Zacharis278 commented on edx_exams/apps/core/models.py:
```diff
@@ -234,6 +234,27 @@ def check_no_other_active_attempts_for_user(cls, user_id, attempt_id):
             return True
 
+    @classmethod
+    def get_attempt_by_attempt_number_and_resource_id(cls, attempt_number, resource_id):
```
nothing wrong with having this here I just want to point out that we don't need always need this type of logic to be on the model class directly. If it's a pretty specific/narrow use case (filter by resource_id feels pretty narrow) it's totally fine to just put in on api.py. We don't always need the extra layer of abstraction.

My rule of thumb here is to do this logic on the model if I expect it would be used multiple times elsewhere.

ilee2u commented:
Also, I'm aware that I haven't reached full coverage. I want to make sure my approach for this ticket is okay first before making tests.

Zacharis278 commented on edx_exams/apps/core/api.py:
```diff
@@ -41,6 +41,17 @@ def get_latest_attempt_for_user(user_id):
     return latest_attempt
 
 
+def get_attempt_by_attempt_number_and_resource_id(attempt_number, resource_id):
```
it looks like we have a very similar function already for `get_attempt_for_user_with_attempt_number_and_resource_id` any way we can cut down on duplicate code paths here?

Zacharis278 commented on edx_exams/apps/lti/views.py:
```diff
@@ -12,19 +15,96 @@
+        ExamAttemptStatus.submitted,
+    ]
+
+    attempt = get_attempt_by_attempt_number_and_resource_id(attempt_number, resource_id)
```
this would also need to look at the user id because attempt_number is not unique. It's always 1 for a user's first attempt, 2 for the second, and so on.

Zacharis278 replied:
What's our error condition if we can't find an attempt for this user and this returns None?

Zacharis278 commented on edx_exams/apps/lti/views.py:
```diff
@@ -12,19 +15,96 @@
+        return Response(status=400)
+
+    if action == 'flag':
+        log.info('Flagging exam attempt')
```
I think it's worth logging the user_id, exam_id, and attempt_id here

Zacharis278 reviewed (commented):
some initial comments

ilee2u commented on edx_exams/apps/core/models.py:
```diff
@@ -241,11 +241,20 @@ def get_attempt_for_user_with_attempt_number_and_resource_id(cls, user_id, attem
-                attempt_number=attempt_number,
-                exam__resource_id=resource_id
-            )
+            # If user_id is an integer, use our internal database ID
```
I'm wondering if we should split this logic between two different model functions, with the if/else statement happening in api.py instead.

Zacharis278 replied:
tbh I wouldn't hide this behavior all the way down here. You can simply get the User in the view `user = User.object.get(anonymous_user_id=<sub_id>)` and than pass user.id down into the existing functions we have. More code reuse and we don't have multiple concepts of what user_id means.

The extra DB call is pretty much irrelevant performance wise. We don't need to over-optimize such that everything is one ORM call.

ilee2u commented:
Small decisions note: the statuses in which an exam attempt can be flagged (ready_to_start, started, ready_to_submit, timed_out, submitted) are subject to change depending on the functionality that our proctoring vendors request in the future.

Zacharis278 commented on edx_exams/apps/core/api.py:
```diff
@@ -48,6 +48,27 @@ def get_latest_attempt_for_user(user_id):
     return latest_attempt
 
 
+def get_user_by_anonymous_id(anonymous_user_id, attempt_number, resource_id):
```
This function name seems to imply this returns the user for an anonymous id but it in fact returns an exam attempt. Overall comment on this is similar to my previous tho. Any way we can shoot for some code reuse here? We already have an api function that gives us an attempt for user, resource_id, and attempt_number which has near identical code.

If this just returned the user we could reuse that existing function based on the id of that user.

Zacharis278 commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -155,9 +387,9 @@ def test_start_proctoring_unauthorized_user(self, mock_get_lti_launch_url):  # p
 @patch('edx_exams.apps.lti.views.get_lti_1p3_launch_start_url', return_value='https://www.example.com')
 class LtiEndAssessmentTestCase(ExamsAPITestCase):
-    """
+    '''
```
Convention across our repos has docstrings with double quotes so I'd recommend we stick with that.

MichaelRoytman commented on edx_exams/apps/lti/views.py:
```diff
@@ -12,19 +15,113 @@
+
 EDX_OAUTH_BACKEND = 'auth_backends.backends.EdXOAuth2'
 
+LTI_PROCTORING_ASSESSMENT_CONTROL_ACTIONS = [
```
You aren't using this constant. Do you still need it? Is it meant to be renamed `VALID_STATUSES`.

If so, keep in mind that you have some mismatches between the values in this list and what the code is looking for. For example, this variable has the value `flagRequest`, but the view code is checking for `flag`.

MichaelRoytman commented on edx_exams/apps/lti/views.py:
```diff
@@ -78,9 +175,16 @@ def start_proctoring(request, attempt_id):
 
+    assessment_control_url = urljoin(
+        get_lti_root(),
+        reverse('lti:acs', kwargs={'lti_config_id': lti_config_id}),  # if the url actually existed
```
Nit: I assume the comment is a note to self, so please remove it.

MichaelRoytman commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
 
 
+@ddt.ddt
+class LtiAcsTestCase(ExamsAPITestCase):
```
What do you think about refactoring this a little to use some of the [factories](https://github.com/edx/edx-exams/blob/main/edx_exams/apps/api/test_utils/factories.py) Zach introduced?

Zacharis278 commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
+        self.assertEqual(response.data, expected_msg)
+        mock_get_attempt.assert_called_once_with(self.user.id, self.attempt.attempt_number, self.exam.resource_id)
+
+    # TEST 3: get_attempt returned None -> Expected log
```
If you want to describe what the test is doing you're better off using a function docstring to do this. The linter should make you do it anyway.

MichaelRoytman replied:
Nit: Please remove this comment.

MichaelRoytman commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
+        }
+
+        # Even though the client.post function below uses json.dumps to serialize the request as json,
+        # The json serialization needs to happen before the request for an unkown reason
```
```suggestion
        # The json serialization needs to happen before the request for an unknown reason
```

MichaelRoytman commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
+                f'Attempt cannot be flagged for user with anonymous id {self.user.anonymous_user_id} '
+                f'with resource id {self.exam.resource_id} and attempt number {self.attempt.attempt_number} '
+                f'for lti config id {self.test_provider.lti_configuration_id}, exam id {self.exam.id}, and '
+                f'attempt id {self.attempt.id}. It has either not started yet, been rejcected, expired, or '
```
If you correct `rejcected` to `rejected`, this will fail, as a heads up.

MichaelRoytman commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
+
+        self.assertEqual(response.status_code, expected_response_status)
+        self.assertEqual(response.data, expected_msg)
+        mock_get_attempt.assert_called_once_with(self.user.id, self.attempt.attempt_number, self.exam.resource_id)
```
I don't know that you need to test this. The most important thing to test is _what_ has happened, not _how_ it happened. I don't think it's important to know that this particular function was called. But please let me know if you disagree!

MichaelRoytman commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
+        }
+
+        # Even though the client.post function below uses json.dumps to serialize the request as json,
+        # The json serialization needs to happen before the request for an unkown reason
```
```suggestion
        # The json serialization needs to happen before the request for an unknown reason
```

Zacharis278 commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
 
 
+@ddt.ddt
+class LtiAcsTestCase(ExamsAPITestCase):
```
we could probably use some limited testing on auth here. It's mostly handled in the library but it's common to have something basic to make sure everything configured properly. For example when we do: https://github.com/edx/edx-exams/blob/e3e375993afe46b0bee6db91b163f5f0ee10a69b/edx_exams/apps/api/v1/tests/test_views.py#L82

Maybe in this case we just check for no token and a token without the correct scope.

ilee2u replied:
I just pushed some code with modifications regarding all of the other comments. I also created a test for when authentication fails.

I tried making a test where the token is valid but doesn't have the correct scope and had some issues (I can go into more detail about this if you'd like). I noticed that we already test the permissions class for the correct scope within xblock-lti-consumer ([Link](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_1p3/tests/extensions/rest_framework/test_permissions.py#L281)), so I'm wondering if there's a reason we should still test for the scope within this repo.

Zacharis278 replied:
ahh ok usually it's pretty easy to test so we do but I see how you'd have to setup an entire config with a non-proctoring consumer which is weird so let's leave this one out.

Zacharis278 commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
+            is_active=True
+        )
+
+        self.attempt = ExamAttempt.objects.create(
```
might want to take a look at the Factory classes added with #129. I think you can do the whole setup of this config/exam/attempt with only a couple lines.

Zacharis278 commented on edx_exams/apps/lti/tests/test_views.py:
```diff
@@ -1,20 +1,248 @@
+        token = self._make_access_token('https://purl.imsglobal.org/spec/lti-ap/scope/control.all')
+
+        # Request w/ attempt number for an attempt that does not exist
+        request_body = {
```
setting up this request body and making the request is the same across both of our tests here. Maybe a helper function to do this? Especially if we're adding another test for auth.

ilee2u commented:
Notes for latest push:

- Streamlined tests to only test for higher level view functionality
- Added a basic authentication failure test
- Implemented the new factories that were added by Zach in a previous PR into my tests
- Other various cleanup items

Zacharis278 reviewed (approved):
looking good! 🚀
